java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/protocol/LspTemplateUI.java

```diff
@@ -0,0 +1,215 @@
+import org.eclipse.lsp4j.WorkspaceFolder;
+import org.openide.filesystems.FileObject;
+import org.openide.filesystems.FileUtil;
+import org.openide.loaders.DataFolder;
+import org.openide.loaders.DataObject;
+import org.openide.loaders.DataObjectNotFoundException;
+import org.openide.nodes.Node;
+import org.openide.util.NbBundle;
+
+class LspTemplateUI {
+    private LspTemplateUI() {
+    }
+
+    static CompletableFuture<Object> createFromTemplate(String templates, NbCodeLanguageClient client, ExecuteCommandParams params) {
+        final FileObject fo = FileUtil.getConfigFile(templates);
+        final DataFolder folder = DataFolder.findFolder(fo);
+        return displayUI(folder, client, params);
+    }
+
+    @NbBundle.Messages({
+        "CTL_TemplateUI_SelectGroup=Select Group of Objects",
+        "CTL_TemplateUI_SelectTemplate=Select Object Template",
+        "CTL_TemplateUI_SelectTarget=Where to put the object?",
+        "CTL_TemplateUI_SelectName=Name of the object?",
+    })
+    private static CompletableFuture<Object> displayUI(DataFolder templates, NbCodeLanguageClient client, ExecuteCommandParams params) {
+        final FileObject[] group = { null };
+        final DataObject[] source = { null };
+        final DataFolder[] target = findTargetFolder(params);
+
+        final List<QuickPickItem> categories = quickPickTemplates(templates);
+        final CompletionStage<List<QuickPickItem>> pickGroup = client.showQuickPick(new ShowQuickPickParams(Bundle.CTL_TemplateUI_SelectGroup(), false, categories));
+        final CompletionStage<List<QuickPickItem>> pickProject = pickGroup.thenCompose(selectedGroups -> {
+            group[0] = templates.getPrimaryFile().getFileObject(singleSelection(selectedGroups));
+            List<QuickPickItem> projectTypes = quickPickTemplates(DataFolder.findFolder(group[0]));
+            return client.showQuickPick(new ShowQuickPickParams(Bundle.CTL_TemplateUI_SelectTemplate(), false, projectTypes));
+        });
+        final CompletionStage<DataObject> findTemplate = pickProject.thenApply((selectedTemplates) -> {
+            try {
+                final String templateName = singleSelection(selectedTemplates);
+                final FileObject templateFo = group[0].getFileObject(templateName);
+                source[0] = DataObject.find(templateFo);
+            } catch (DataObjectNotFoundException ex) {
+                throw raise(RuntimeException.class, ex);
+            }
+            return source[0];
+        });
+        final CompletionStage<DataFolder> findTarget;
+        if (target[0] == null) {
+            findTarget = findTemplate.thenCompose(any -> client.workspaceFolders()).thenCompose(folders -> {
+                return client.showInputBox(new ShowInputBoxParams(Bundle.CTL_TemplateUI_SelectTarget(), findWorkspaceRoot(folders)));
+            }).thenApply((path) -> {
+                target[0] = DataFolder.findFolder(FileUtil.toFileObject(new File(path)));
```

Comment:
The target here may be something completely screwed up, or even empty. Are error messages from `DataFolder.find` reasonable to display in LSP client so the user understands his/her mistake ?

Reply:
Entered path is now validated since 819c639 - I hope you'll appreciate the recursive `CompletionStage` created when it is not valid.